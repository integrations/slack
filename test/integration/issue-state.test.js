const supertest = require('supertest');
const nock = require('nock');
const { promisify } = require('util');

const { probot, slackbot, models } = require('.');
const fixtures = require('../fixtures');

const request = supertest.agent(probot.server);

const { SlackWorkspace, SlackUser, GitHubUser } = models;

const sleep = promisify(setTimeout);

describe('Integration: issue state', () => {
  let user;
  let workspace;

  beforeEach(async () => {
    user = await GitHubUser.create({
      id: 2,
      accessToken: 'github-token',
    });

    workspace = await SlackWorkspace.create({
      slackId: 'T0001',
      accessToken: 'xoxa-token',
    });

    await SlackUser.create({
      slackId: 'U2147483697',
      slackWorkspaceId: workspace.id,
      githubId: user.id,
    });
  });

  test('/github close issue', async () => {
    nock('https://api.github.com').get('/repos/owner/repo/installation')
      .reply(200, { ...fixtures.installation, account: fixtures.repo.owner });
    nock('https://api.github.com')
      .get('/repos/owner/repo')
      .reply(200, fixtures.repo);
    nock('https://api.github.com')
      .get('/user')
      .reply(200, fixtures.user);
    nock('https://api.github.com')
      .patch('/repos/owner/repo/issues/123', { state: 'closed' })
      .reply(200);

    nock('https://slack.com').post('/api/chat.postMessage', (body) => {
      expect(JSON.stringify(body)).toMatch(/Issue closed by wilhelmklopp/);
      return true;
    }).reply(200, { ok: true });

    const command = fixtures.slack.command({
      text: 'close https://github.com/owner/repo/issues/123',
    });

    await request.post('/slack/command').use(slackbot).send(command)
      .expect(200, { response_type: 'in_channel' });

    // wait for request to post to slack after command response
    await sleep(10);
  });

  test('/github reopen issue', async () => {
    nock('https://api.github.com').get('/repos/owner/repo/installation')
      .reply(200, { ...fixtures.installation, account: fixtures.repo.owner });
    nock('https://api.github.com')
      .get('/repos/owner/repo')
      .reply(200, fixtures.repo);
    nock('https://api.github.com')
      .get('/user')
      .reply(200, fixtures.user);
    nock('https://api.github.com')
      .patch('/repos/owner/repo/issues/123', { state: 'open' })
      .reply(200);

    nock('https://slack.com').post('/api/chat.postMessage', (body) => {
      expect(JSON.stringify(body)).toMatch(/Issue reopened by wilhelmklopp/);
      return true;
    }).reply(200, { ok: true });

    const command = fixtures.slack.command({
      text: 'reopen https://github.com/owner/repo/issues/123',
    });

    await request.post('/slack/command').use(slackbot).send(command)
      .expect(200, { response_type: 'in_channel' });

    // wait for request to post to slack after command response
    await sleep(10);
  });

  test('missing permissions', async () => {
    nock('https://api.github.com').get('/repos/owner/repo/installation')
      .reply(200, {
        ...fixtures.installation,
        account: fixtures.repo.owner,
        permissions: {
          ...fixtures.installation.permissions,
          pull_requests: 'write',
          issues: 'read',
        },
      });

    nock('https://api.github.com').get('/repos/owner/repo')
      .reply(200, fixtures.repo);

    nock('https://api.github.com').get('/user')
      .reply(200, fixtures.user);

    nock('https://api.github.com')
      .patch('/repos/owner/repo/issues/123', { state: 'closed' })
      .reply(403);

    const command = fixtures.slack.command({
      text: 'close https://github.com/owner/repo/issues/123',
    });


    await request.post('/slack/command').use(slackbot).send(command)
      .expect(200, /requires updated permissions/);
  });
});
