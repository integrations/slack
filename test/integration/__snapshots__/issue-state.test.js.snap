// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: issue state /github close issue 1`] = `
Array [
  "channel#1#C2147483705:issue#undefined",
  "pending-command#13345224609.738474920.8088930838d88f008e0",
]
`;

exports[`Integration: issue state /github close issue 2`] = `
Array [
  Array [
    "channel#1#C2147483705:issue#undefined",
    Object {
      "channel": "C2147483705",
      "ts": undefined,
    },
  ],
  Array [
    "pending-command#13345224609.738474920.8088930838d88f008e0",
    Object {
      "channel_id": "C2147483705",
      "channel_name": "test",
      "command": "/github",
      "enterprise_id": "E0001",
      "enterprise_name": "Globular%20Construct%20Inc",
      "response_url": "https://hooks.slack.com/commands/1234/5678",
      "team_domain": "example",
      "team_id": "T0001",
      "text": "close https://github.com/owner/repo/issues/123",
      "token": "secret",
      "trigger_id": "13345224609.738474920.8088930838d88f008e0",
      "user_id": "U2147483697",
      "user_name": "Steve",
    },
  ],
]
`;

exports[`Integration: issue state /github reopen issue 1`] = `
Array [
  "channel#1#C2147483705:issue#undefined",
  "pending-command#13345224609.738474920.8088930838d88f008e0",
]
`;

exports[`Integration: issue state /github reopen issue 2`] = `
Array [
  Array [
    "channel#1#C2147483705:issue#undefined",
    Object {
      "channel": "C2147483705",
      "ts": undefined,
    },
  ],
  Array [
    "pending-command#13345224609.738474920.8088930838d88f008e0",
    Object {
      "channel_id": "C2147483705",
      "channel_name": "test",
      "command": "/github",
      "enterprise_id": "E0001",
      "enterprise_name": "Globular%20Construct%20Inc",
      "response_url": "https://hooks.slack.com/commands/1234/5678",
      "team_domain": "example",
      "team_id": "T0001",
      "text": "reopen https://github.com/owner/repo/issues/123",
      "token": "secret",
      "trigger_id": "13345224609.738474920.8088930838d88f008e0",
      "user_id": "U2147483697",
      "user_name": "Steve",
    },
  ],
]
`;

exports[`Integration: issue state missing permissions 1`] = `
Array [
  "pending-command#13345224609.738474920.8088930838d88f008e0",
]
`;

exports[`Integration: issue state missing permissions 2`] = `
Array [
  Array [
    "pending-command#13345224609.738474920.8088930838d88f008e0",
    Object {
      "channel_id": "C2147483705",
      "channel_name": "test",
      "command": "/github",
      "enterprise_id": "E0001",
      "enterprise_name": "Globular%20Construct%20Inc",
      "response_url": "https://hooks.slack.com/commands/1234/5678",
      "team_domain": "example",
      "team_id": "T0001",
      "text": "close https://github.com/owner/repo/issues/123",
      "token": "secret",
      "trigger_id": "13345224609.738474920.8088930838d88f008e0",
      "user_id": "U2147483697",
      "user_name": "Steve",
    },
  ],
]
`;
